pkg/sentry/fsimpl/fuse/dev_test.go

```diff
@@ -0,0 +1,373 @@
+				serverChans[j] = make(chan struct{})
+				serverKillChans[j] = make(chan struct{}, 1) // The kill command shouldn't block.
+				go func(j int) {
+					fuseServerRun(t, s, k, fd, serverChans[j], serverKillChans[j])
+				}(j)
+			}
+
+			// Make sure all the clients are done.
+			for i := 0; i < testCase.NumClients; i++ {
+				<-clientChans[i]
+			}
+
+			// Kill any server that is potentially waiting.
+			for j := 0; j < testCase.NumServers; j++ {
+				serverKillChans[j] <- struct{}{}
+			}
+
+			// Make sure all the servers are done.
+			for j := 0; j < testCase.NumServers; j++ {
+				<-serverChans[j]
+			}
+		})
+	}
+}
+
+// CallTest makes a request to the server and blocks the invoking
+// goroutine until a server responds with a response. Doesn't block
+// a kernel.Task. Analogous to Connection.Call but used for testing.
+func CallTest(conn *Connection, r *Request) (*Response, error) {
+	fut, err := conn.callFuture(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// Block without a task.
+	select {
+	case <-fut.ch:
+	}
+
+	// A response is ready. Resolve and return it.
+	return fut.getResponse(), nil
+}
+
+// ReadTest is analogous to vfs.FileDescription.Read and reads from the FUSE
+// device. However, it does so by - not blocking the task that is calling - and
+// instead just waits on a channel. The behaviour is essentially the same as
+// DeviceFD.Read except it guarantees that the task is not blocked.
+func ReadTest(serverTask *kernel.Task, fd *vfs.FileDescription, inIOseq usermem.IOSequence, killServer chan struct{}) (int64, bool, error) {
+
+	// A lock is needed to guarantee that if a server is woken up (a request is
+	// available), then that server will process the request. We can't have another
+	// server running in parallel barge in and serve the request as that would cause
+	// the serverTask to block when Read is called below. This is needed only during
+	// testing - when actually running FUSE in gVisor, we don't need to avoid
+	// blocking the task.
+	testReadMu.Lock()
+	defer testReadMu.Unlock()
+
+	// Emulate the blocking for when no requests are available. We can't really
+	// block a task during test and so this way we guarantee the fast path of
+	// task.Block() (when no wait is required) is hit.
+	waitChan := fd.Impl().(*DeviceFD).waitCh
+	select {
+	case <-waitChan:
+		// Make sure there is something for Read to find.
+		waitChan <- struct{}{}
+	case <-killServer:
+		// Server killed by the main program.
+		return 0, true, nil
+	}
+
+	// Perform a non blocking read.
+	n, err := fd.Read(serverTask, inIOseq, vfs.ReadOptions{})
+	return n, false, err
+}
+
+// fuseClientRun emulates all the actions of a normal FUSE request. It creates
+// a header, a payload, calls the server, waits for the response, and processes
+// the response.
+func fuseClientRun(t *testing.T, fuseConn *Connection, creds *auth.Credentials, pid uint32, inode uint64, clientDone chan struct{}) {
+	testObj := &testObject{
+		opcode: testOpcode,
```

Comment:
Perhaps randomizing this would be a good idea? Will catch potential bugs like when the server misused the optcode from the previous one.